Add database integration docs (#808)
kubearchive/kubearchive: docs/modules/ROOT/pages/integrations/database.adoc

= Database Integrations

== Overview

KubeArchive requires a database to store resources.
This document list the databases supported, how to set up the database schema
and how to add new database integrations.

=== PostgreSQL

> The only currently full supported.

Reviewer: ```suggestion
Currently PostgreSQL is the only fully supported.
```

Author: Applied the changes with a little modification:

`Currently, PostgreSQL is the only fully supported database engine.`

WDYT?

> KubeArchive integration test suite runs with PostgreSQL 16.6

Reviewer: ```suggestion
The KubeArchive integration test suite runs with PostgreSQL 16.6
```

Author: applied suggestion

The PostgreSQL implementation is available
link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/postgresql.go[here].

> 

Reviewer: The PostgreSQL schema is available

link:https://github.com/kubearchive/kubearchive/blob/main/integrations/database/postgresql/kubearchive.sql[here]
and
link:https://github.com/kubearchive/kubearchive/releases/latest[the latest release as `kubearchive.sql`].

Author: applied suggestion

=== MariaDB

This integration is still in progress.

The MariaDB implementation is available
link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/mariadb.go[here].

== Configuration and Customization

KubeArchive offers a [filename]`kubearchive.sql` file with the DDL to create
the schema as part of the installation.

[IMPORTANT]
====
The provided schema includes indexes.

KubeArchive users should review and customize the indexes
if needed based on the expected queries and the amount and

> the distribution of the expected archived data.

Author: suggestion applied!

====

Specific instructions for the configuration of the PostgreSQL Database are available in

> xref:/getting-started/instllation.adoc#_install_and_configure_kubearchive[Install and Configure KubeArchive].

Author: From the user POV, I think it's interesting having also instructions (or a link to them in this case) with the instructions of how "install" the database, as it's not only applying the schema, the user need to know about how the credentials are exposed.

> xref:/getting-started/instllation.adoc#_install_and_configure_kubearchive[Install and Configure KubeArchive].

Reviewer: I agree, maybe we need a section about the configuration in the secret and then link it to the specific example given in the getting started pages

> xref:/getting-started/instllation.adoc#_install_and_configure_kubearchive[Install and Configure KubeArchive].

Author: one way or another works for me as long as the information isn't repeated (we should have links, this makes the doc more maintainable) and the information is in both db integration and getting started sections.

If you have a strong preference of having the instructions of configuring the secret in the "database integration" section, I can add the changes.

> xref:/getting-started/instllation.adoc#_install_and_configure_kubearchive[Install and Configure KubeArchive].

Reviewer: I don't agree that the information here and in the getting started are the same. This document feels as a reference guide (definition + brief explanation), while the getting started is more of a tutorial (don't understand, just apply and get something working). So here it makes sense that we explain the secret and its keys, while it does not make sense to have a `kubectl patch/apply` example. So explain the secret keys here and link to the getting started as an example.

> xref:/getting-started/instllation.adoc#_install_and_configure_kubearchive[Install and Configure KubeArchive].

Author: I've added some instructions regarding how to apply the schema and the patch to the `kubearchive-database-credentials` Secret.
PTAL

> == Add new database integrations

Reviewer: ```suggestion
== Adding A New Database
```

Author: suggestion applied

To add a new database integration follow the instructions:

> 1. Create a new file under `pkg/database` with the name of the new database type.

Reviewer: ```suggestion
1. Create a new file under `pkg/database` named after the new database engine.
```

Author: suggestion applied

2. Include an
link:https://go.dev/doc/effective_go#init[init function]
that:

* Inserts in the

> link:https://github.com/maruiz93/kubearchive/blob/postgresql-version-update/pkg/database/database.go#L27[`RegisteredDBCreators`]

Reviewer: Should this point to our main repository instead?

```suggestion
link:https://github.com/kubearchive/kubearchive/blob/postgresql-version-update/pkg/database/database.go#L27[`RegisteredDBCreators`]
```

Author: ops, yes, it should

Author: fixed

map an entry keyed with the name of the DB type with a
link:https://github.com/maruiz93/kubearchive/blob/postgresql-version-update/pkg/database/database.go#L24[`newDBCreatorFunc`]
implementation as value.

* Inserts in the
link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/database.go#L26[`RegisteredDatabases`]
map an entry keyed with the name of the DB type with a
link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/database.go#L23[`newDatabaseFunc`]
as value.

=== The `newDBCreatorFunc`

This function receives the set of environment variables with the database configuration and returns a
link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/facade/creator.go[`DBCreator`]
implementation.

> The recommendation is that the `DBCreator` implementation include the `env` param as parameter
> so the `GetDriverName` and the `GetConnectionString` can access them.

Reviewer: ```suggestion
The `DBCreator` implementation must include the `env` param as parameter
so the `GetDriverName` and the `GetConnectionString` can access them.
```

Author: I've replaced it with a `should`.

You can potentially do a working `GetConnectionString` method that doesn't depend on an `env`, but you shouldn't do that because that breaks our "development guidelines".

Reviewer: I'm fine with should instead of must. However I think this should not be about what you could do, but about what we want you to do as maintainers of the tool.

=== The `DBInterface` implementation returned by the `newDatabaseFunc`

[NOTE]
====
This section includes the development guidelines expected by the KubeArchive maintainers.
They can be skipped, but they will be considered in the review process.

> ====

Reviewer: If they are optional but they will be considered in the review process they are in fact required.

Author: I removed the second sentence.

The reason I added the note is because these guidelines may not work for every database engine. For example, when adding integration with a non-relational db like mongo.

There might be other examples, but the point is, we will review the PRs considering these guidelines whenever they can be applied.

The `DBInterface` should be a struct with
link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/database.go#L40[`Database`]
embedded.

Make sure to implement the interfaces that are part of it:

* link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/facade/deleter.go[DBDeleter]
* link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/facade/filter.go[DBFilter]
* link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/facade/inserter.go[DBInserter]
* link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/facade/selector.go[DBSelector]
* link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/facade/sorter.go[DBSorter]

[NOTE]
====
All the interfaces work with objects from the
link:https://pkg.go.dev/github.com/huandu/go-sqlbuilder[sqlbuilder Go library].
Check out the docs to see how to work with them.
====

[TIP]
====
Some interfaces, like `DBDeleter`, have a default implementation.

Other interfaces, like `DBFilter`, have a partial implementation.

Those implementations can have the functionality that you need.

> Check them before implementing your own.

Author: suggestion applied

====

> If the database interaction logic changes from the implemented in the `Database` struct,
> override the implementation of their functions.

Reviewer 2: ```suggestion
If the database interaction logic changes from the implementation in the `Database` struct,
override the implementation of the affected functions.
```

Author: suggestion applied

> If the database interaction logic changes from the implemented in the `Database` struct,
> override the implementation of their functions.
> An example could be with a database that doesn't support upsert queries
> and need to do a transaction with a `SELECT` and an `INSERT`.

Reviewer: ```suggestion
If the database interaction logic changes from the implemented in the `Database` struct,
override the implementation of their functions. For example, a database that does not
support upsert queries needs to implement `ResourceInserter` as a series of  `SELECT` and an `INSERT` statements.
```

Author: no, currently neither mariadb or postgresql reimplement those methods, they just use the "default"implementation of the `Database` struct as it is embedded there. 
It's not correct as go doesn't support the inheritance, it uses composition, but the outcome there is like both `PostgreSQLDatabase` and `MariaDBDatabase` "inherit" the methods from `Database` and they don't need to override them.

= Database Integrations

== Overview

KubeArchive requires a database to store resources.
This document list the databases supported, how to set up the database schema
and how to add new database integrations.

=== PostgreSQL

Currently, PostgreSQL is the only fully supported database engine.

> The KubeArchive integration test suite runs with PostgreSQL 16.6

Author: > Can we just point to the kubearchive.sql file to tell the user which version is supported? Then we don't have to change it in multiple places.

It's ok for me, I'll do the change

> Also, why are we talking about the integration test suite? Shouldn't it just say KubeArchive needs PostgreSQL version of 16.6 or later?

Hmm, I don't think so, I'm pretty sure that KubeArchive works with lower versions of PostgreSQL, it's just that we test the product with a specific one, so if the user wants to try a different one it's more "risky" but it's probably going to be ok unless they use a really old version. Maybe a couple of errors on the import of the pg_dump, but as far as I'm concerned, related to the queries we don't use any really recent functionality.

> The KubeArchive integration test suite runs with PostgreSQL 16.6

Reviewer 3: > > Also, why are we talking about the integration test suite? Shouldn't it just say KubeArchive needs PostgreSQL version of 16.6 or later?
> 
> Hmm, I don't think so

Typically database version are not backward compatible, but are forward compatible. In other words, if we say we require 16.6, I should not expect < 16.6 to work but would expect >= 16.6 to work. We should say "at least 16.6" or "16.6 or later".

And it's still not clear why we are talking about the integration test suite. KubeArchive needs a particular PostgreSQL version, not the test suite.

> The KubeArchive integration test suite runs with PostgreSQL 16.6

Author: updated

The PostgreSQL implementation is available
link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/postgresql.go[here].

The PostgreSQL schema is available
link:https://github.com/kubearchive/kubearchive/blob/main/integrations/database/postgresql/kubearchive.sql[here]
and
link:https://github.com/kubearchive/kubearchive/releases/latest[the latest release as `kubearchive.sql`].

=== MariaDB

This integration is still in progress.

The MariaDB implementation is available
link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/mariadb.go[here].

== Configuration and Customization

> KubeArchive offers a [filename]`kubearchive.sql` file with the DDL to create

Author: In this sentence I use both `DDL` and `schema` because they are not exactly the same. DDL are the instructions used to defined a schema. Anyway, I get the point that there is no need to add acronyms and people that aren't used to database stuff may not be familiar with them, so I'll rephrase the sentence.

> KubeArchive offers a [filename]`kubearchive.sql` file with the DDL to create

Author: Changed!

link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/mariadb.go[here].

[#_configuration_and_customization]
== Configuration and Customization

=== Schema

KubeArchive offers a [filename]`kubearchive.sql` file with the DDL to create
the schema as part of the installation.

[IMPORTANT]
====
The provided schema includes indexes.

KubeArchive users should review and customize the indexes
if needed based on the expected queries and the amount and
distribution of the expected archived data.
====

The schema includes the creation of the database `kubearchive` owned by the `kubearchive` user.
The schema expects the existence of a `kubearchive` database user fow owning this database.

[IMPORTANT]
====
An account with admin privileges should run [filename]`kubearchive.sql`.

[source, bash]
----
psql -U admin \ <1>
-h database.example.com \ <2>
-f kubearchive.sql
----
<1> The database admin account used to apply the schema. It can be different from the KubeArchive DB user.
<2> The database URL
====

=== Database Credentials

The `kubearchive-database-credentials` Secret stores the information to connect KubeArchive with the Database.

> The default content of this Secret when KubeArchive is installed has the following data:

Reviewer: ```suggestion
The default content of this Secret when KubeArchive is installed is:
```

[source, yaml]
----
kind: Secret
type: Opaque
metadata:
  name: kubearchive-database-credentials
  namespace: kubearchive
stringData:
  DATABASE_KIND: postgresql
  DATABASE_PORT: "5432"
  DATABASE_URL: "kubearchive-rw.postgresql.svc.cluster.local"
  DATABASE_DB: "kubearchive"
  DATABASE_USER: "kubearchive"
  DATABASE_PASSWORD: "Databas3Passw0rd"  # notsecret
----

The KubeArchive user should update the secret with the specific values of their database and
restart the pods accordingly to pick the new values.

> The command for changing the most common values, URL and password, is:

Author: suggestion applied

= Database Integrations

== Overview

> KubeArchive needs a database to store the resources to make them retrievable.
> KubeArchive currently supports integration with PostgreSQL.
> MariaDB integration is in progress.
> 
> [NOTE]
> ====
> 
> KubeArchive integration test suite runs with PostgreSQL 16.6
> 
> ====

Reviewer: ```suggestion
KubeArchive requires a database to store resources. This document list the databases supported, how to setup the database schema and how to configure KubeArchive to connect to the database.
```

Then I would create a section per DB: PostgreSQL, MariaDB. Each section should have the version, as you added in a NOTE below, and then the link to the SQL file.

What do you think?

Author: I've changed the docs to go with this approach, take a look and see if it looks good to you

== Configuration and Customization

KubeArchives offers a [filename]`kubearchive.sql` file with the DDL to create
the schema as part of the installation.

[IMPORTANT]
====
The provided schema includes indexes.

KubeArchive users should review and customize the indexes
if needed based on the expected queries and the amount and
the distribution of the expected archived data.
====

Specific instructions for the configuration of the PostgreSQL Database are available in

> xref:/getting-started/configuration.adoc#_install_and_configure_kubearchive[Install and Configure KubeArchive].

Reviewer: The instructions are not in the `configuration.adoc` document, link directly to the `installation.adoc` file.

Author: Right! It's fixed now

== Support new databases

> KubeArchive database integration is designed to include new databases.

Reviewer: I think this phrase can go away, this idea is already implicit in the section itself.

> To add a new database integration the following interfaces need to be implemented:

Reviewer: ```suggestion
To add support for a new database, implement the following interfaces:
```

Author: applied a similar change to remove the passive form

* link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/facade/creator.go[DBCreator]
* link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/facade/deleter.go[DBDeleter]
* link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/facade/filter.go[DBFilter]
* link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/facade/inserter.go[DBInserter]
* link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/facade/selector.go[DBSelector]
* link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/facade/sorter.go[DBSorter]

[NOTE]
====

> Except for the `DBCreator` functions, all the functions in the interfaces receive/output objects of the
> link:https://pkg.go.dev/github.com/huandu/go-sqlbuilder[sqlbuilder Go library].
> Check out the docs to see how to work with them.

Reviewer: ```suggestion
All the interfaces, with the exception of `DBCreator`, work with objects from the
link:https://pkg.go.dev/github.com/huandu/go-sqlbuilder[sqlbuilder Go library].
Check out the docs to see how to work with them.
```

Author: The sentence was incorrect, I change the instructions so now there is no "except" in the sentence.

====

[TIP]
====

> Some of the interfaces like `DBDeleter` include a default implementation that can be used called `DBDeleterImpl`.
> 
> Other interfaces like `DBFilter` include a partial implementation called `PartialDBFilterImpl` with the
> implementations of the filters that are likely to be reused.
> 
> Check if those default implementations fit your database integration needs before implementing them.

Reviewer: ```suggestion
Some interfaces, like `DBDeleter`, have a default implementation. Other interfaces, like `DBFilter`, have a partial implementation. Those implementations can be useful, check them before implementing your own.
```

====

If more customization is needed, override the implementation of the functions declared in

> link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/database.go#L29[DBInterface].

Reviewer: Which kind of "more customization" are we talking about?

The PostgreSQL implementation is available in

> link:https://github.com/kubearchive/kubearchive/blob/main/pkg/database/postgresql.go[here].

Reviewer: If you liked the idea of having a section per database, this should go into the PostgreSQL section

Author: done!
